src/java.base/share/classes/java/lang/template/TemplateRuntime.java

```diff
@@ -0,0 +1,392 @@
+         * @param name      method name
+         * @param type      method type
+         * @param fragments fragments from string template
+         * @param processor static final processor
+         */
+        private TemplateBootstrap(MethodHandles.Lookup lookup, String name, MethodType type,
+                                  List<String> fragments,
+                                  ValidatingProcessor<?, ? extends Throwable> processor) {
+            this.lookup = lookup;
+            this.name = name;
+            this.type = type;
+            this.fragments = fragments;
+            this.processor = processor;
+
+        }
+
+        /**
+         * Create callsite to invoke specialized processor process method.
+         *
+         * @return {@link CallSite} for processing templated strings.
+         * @throws Throwable if linkage fails
+         */
+        private CallSite processWithProcessor() throws Throwable {
+            MethodHandle mh = processor instanceof ProcessorLinkage processorLinkage ?
+                    processorLinkage.linkage(fragments, type) : defaultProcessMethodHandle();
+
+            return new ConstantCallSite(mh);
+        }
+
+        /**
+         * Creates a simple {@link StringTemplate} and then invokes the processor's process method.
+         *
+         * @param fragments fragments from string template
+         * @param processor {@link ValidatingProcessor} to process
+         * @param values    array of expression values
+         * @return
+         */
+        private static Object defaultProcess(List<String> fragments,
+                                             ValidatingProcessor<Object, Throwable> processor,
+                                             Object[] values) throws Throwable {
+            return processor.process(new SimpleStringTemplate(fragments, List.of(values)));
+        }
+
+        /**
+         * Generate a {@link MethodHandle} which is effectively invokes
+         * {@code processor.process(new StringTemplate(fragments, values...)}.
+         *
+         * @return default process {@link MethodHandle}
+         */
+        private MethodHandle defaultProcessMethodHandle() {
+            MethodHandle mh = MethodHandles.insertArguments(DEFAULT_PROCESS_MH, 0, fragments, processor);
+            mh = mh.withVarargs(true);
+            mh = mh.asType(type);
+
+            return mh;
+        }
+
+    }
+
+    /**
+     * Collect nullable elements from an array into a unmodifiable list.
+     *
+     * @param elements  elements to place in list
+     *
+     * @return unmodifiable list.
+     *
+     * @param <E>  type of elements
+     *
+     * @implNote Intended for use by {@link StringTemplate} implementations.
+     * Other usage may lead to undesired effects.
+     */
+    @SuppressWarnings("unchecked")
+    public static <E> List<E> toList(E... elements) {
+        return Collections.unmodifiableList(Arrays.asList(elements));
```

Comment:
This is List.of(), please use List.of() instead

Reply:
`List.of()` can't be used here, since the elements are nullable, according to the documentation. But the the returned list can still be modified, by changing the given `elements` array. The input array must be explicitly copied:
```java
public static <E> List<E> toList(E... elements) {
	return Collections.unmodifiableList(new ArrayList<>(Arrays.asList(elements)));
}
```

Reply:
Yes, it only occurs to me mid review, that said there is already an implementation in the jdk of a compact immutable that allow null inside the JDK (this implementation is used when stream.toList() is used).
Using that implementation will avoid a bunch of indirection

Reply:
The change to JUCA means the suppress warnings will need to be extended to @SuppressWarnings({"unchecked", "varargs"}).